From 1d4454a088b5cd2df3d0646a7de4a8acaf63586c Mon Sep 17 00:00:00 2001
From: =?utf-8?q?Tommi=20My=C3=B6h=C3=A4nen?= <ext-tommi.1.myohanen@nokia.com>
Date: Fri, 31 Jul 2009 11:57:12 +0300
Subject: [PATCH] camerabin: reset active-pad in output-selector after READY state

Camerabin sets itself to READY state during resolution change. This
operation makes output-selector to forget its currently active pad,
so it must be set again after state change.
---
 gst/camerabin/gstcamerabin.c |   13 +++++++++++++
 1 files changed, 13 insertions(+), 0 deletions(-)

diff --git a/gst/camerabin/gstcamerabin.c b/gst/camerabin/gstcamerabin.c
index 4cfb7b6..b9ddace 100644
--- a/gst/camerabin/gstcamerabin.c
+++ b/gst/camerabin/gstcamerabin.c
@@ -3228,6 +3228,7 @@ gst_camerabin_user_res_fps (GstCameraBin * camera, gint width, gint height,
     gint fps_n, gint fps_d)
 {
   GstState state, pending;
+  GstPad *activepad;
 
   GST_INFO_OBJECT (camera, "switching resolution to %dx%d and fps to %d/%d",
       width, height, fps_n, fps_d);
@@ -3235,6 +3236,8 @@ gst_camerabin_user_res_fps (GstCameraBin * camera, gint width, gint height,
   /* Interrupt ongoing capture */
   gst_camerabin_do_stop (camera);
 
+  g_object_get (G_OBJECT (camera->src_out_sel), "active-pad", &activepad, NULL);
+
   gst_element_get_state (GST_ELEMENT (camera), &state, &pending, 0);
   if (state == GST_STATE_PAUSED || state == GST_STATE_PLAYING) {
     GST_INFO_OBJECT (camera,
@@ -3250,6 +3253,16 @@ gst_camerabin_user_res_fps (GstCameraBin * camera, gint width, gint height,
         gst_element_state_get_name (pending));
     state = pending;
   }
+
+  /* Re-set the active pad since switching camerabin to READY state clears this
+   * setting in output-selector */
+  if (activepad) {
+    GST_INFO_OBJECT (camera, "re-setting active pad in output-selector");
+
+    g_object_set (G_OBJECT (camera->src_out_sel), "active-pad", activepad,
+        NULL);
+  }
+
   gst_element_set_state (GST_ELEMENT (camera), state);
 }
 
